### DISCUSSION TITLE
NavigationStack(path:root:destination:) not rendering root view when scoping the store into a new one.

### DISCUSSION THREAD

**Discussion by danielsan22** [pointfreeco/swift-composable-architecture#2958] [General]
### Description

Hi all, first of all, great work on this library! 

I wanted to implement navigation within the app I'm working on to learn/practice more about TCA, I implemented a new feature `RootFeature` that will hold the state of the stack, and added there the initial state of my top most feature to then scope it out and have it as the root of the `NavigationStack(path:root:destination)` view. 
But when building the main feature using the RootFeature I only get a white screen, that does loads and sends the initial `.onTask` action but never updates anything else, even when printing the changes on the reducer I can see that there is no change on the state and no other action is sent even tho the `.onTask` sends an array of data.

However, if I create a new store and pass it to the MainView inside the root parameter of the NavigationStack I do get the screen displaying and all the functionality that I expect, but I haven't found a way of comunicating to the parent `RootFeature` and get the actions that then will trigger the navigation I have defined on the `RootFeature`.

I am unsure if this is something that I should be doing or not or instead focus on having all the navigation logic directly inside the feature that will be the "true" root feature, but I like having my features a bit more isolated, thats why having the RootFeature take care of all the details of the other screens/features makes sense to me, it is also how I understood the docs since the  examples use a simple NavigationLink as root (which to my app does not make sense to do that).

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

The NavigationStack's `root` view can be created using a scoped version of the `RootFeature'`s store and renders and works with all the defined actions.

### Actual behavior

The NavigationStack's `root` view is not displayed and the console prints out no changes to the sate even when they are hard-coded as when the view loads.

### Steps to reproduce

I have created a [brand new project](https://github.com/danielsan22/NavStackReproduction) and recreated the issue there.

### The Composable Architecture version information

1.9.2

### Destination operating system

iOS 17

### Xcode version information

15.0

### Swift Compiler version information

```shell
swift-driver version: 1.82.2 Apple Swift version 5.9 (swiftlang-5.9.0.114.10 clang-1500.0.29.1)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @danielsan22, the problem is in the `body` of the `RootFeature`:

https://github.com/danielsan22/NavStackReproduction/blob/973142a5ef7a4e7153d7fbe2a6ef97bf7d315256/NavStackReproduction/RootView.swift#L19-L31

You aren't composing the `MainFeature` reducer into this at all. It should be more like this:

```diff
 var body: some ReducerOf<Self> {
+    Scope(state: \.cardsSets, action: \.cardsSets) {
+      MainFeature()
+    }
     Reduce { _, action in
         switch action {
         case .path:
             return .none
         case .cardSets:
             return .none
         }
     }
     .forEach(\.path, action: \.path) {
         Path()
     }
 }
```

**Comment by mbrandonw** [comment]
Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue to discuss over there.

**Comment by danielsan22** [comment]
Ah that was it, I was under the impresion that I only needed that for the views that where going to be on the stack. 
Thanks for the help!

**Comment by danielsan22** [comment]
I have a follow up question about stack based navigation:
When using stack-based navigation, how can a deeply nested feature (FeatureB) communicate with the previous feature that invoke it (FeatureA)?
The current approach is to listen to child events on the RootFeature and then push the feature to the stack, but there is no way to listen for child features inside a child feature
eg: I have the following feature stack in the StackState: RootFeature -> MainFeature -> FeatureA -> FeatureB.

What I would like to do is to get the actions of `FeatureB` inside `FeatureA` to update accordingly, which I don't know how to do because the navigation is all driven by the RootFeature.

~~I also tried to use a `NavigationLink` inside `FeatureA`, to have `FeatureB.`State as an optional state and therefore be able to listen to the actions of `FeatureB`, but I haven't found an overload that will allow me to push it to the RootFeature path, which I feel like can lead to some weird issues in the future.~~
EDIT: Digging deeper rin the Standups Series's examples I found the way to push directly from the view. So the only question would be how to listen for the updates on a more dynamic way, as for that I haven't found an example yet.

**Comment by danielsan22** [comment]
I've seen this on the Stanups series, listening to the delegate of the deeply nested feature and then using something like `let detailID = state.path.ids.dropLast().last` to get the previous state and perform the changes we need there, but I was wondering if there is a more dynamic way of handling this, as the solution used on the Standups App is assuming the previously pushed feature is a specific one instead of allowing whatever feature wants to push a feature and act upon its actions.

**Reply by mbrandonw** [reply]
There are just tradeoffs to using a fully statically typed navigation stack versus a dynamic one. If you want to use a statically typed one, as we do in SyncUps, then you have to do some extra work to analyze each feature in the stack and see what is going on. I can't provide any general guidance for you other than the `StackAction` and `StackState` hold all of the information you need to figure things out.
